src/Utilities/penalty_relaxation.jl

```diff
@@ -0,0 +1,310 @@
+
+It does not include variable bound or integrality constraints, because these
+cannot be modified in-place.
+
+To modify variable bounds, rewrite them as linear constraints.
+
+## Examples
+
+```jldoctest; setup=:(import MathOptInterface; const MOI = MathOptInterface)
+julia> model = MOI.Utilities.Model{Float64}();
+
+julia> x = MOI.add_variable(model);
+
+julia> c = MOI.add_constraint(model, 1.0 * x, MOI.LessThan(2.0));
+
+julia> map = MOI.modify(model, MOI.Utilities.PenaltyRelaxation(default = 2.0));
+
+julia> print(model)
+Minimize ScalarAffineFunction{Float64}:
+ 0.0 + 2.0 v[2]
+
+Subject to:
+
+ScalarAffineFunction{Float64}-in-LessThan{Float64}
+ 0.0 + 1.0 v[1] - 1.0 v[2] <= 2.0
+
+VariableIndex-in-GreaterThan{Float64}
+ v[2] >= 0.0
+
+julia> map[c] isa MOI.ScalarAffineFunction{Float64}
+true
+```
+
+```jldoctest; setup=:(import MathOptInterface; const MOI = MathOptInterface)
+julia> model = MOI.Utilities.Model{Float64}();
+
+julia> x = MOI.add_variable(model);
+
+julia> c = MOI.add_constraint(model, 1.0 * x, MOI.LessThan(2.0));
+
+julia> map = MOI.modify(model, MOI.Utilities.PenaltyRelaxation(Dict(c => 3.0)));
+
+julia> print(model)
+Minimize ScalarAffineFunction{Float64}:
+ 0.0 + 3.0 v[2]
+
+Subject to:
+
+ScalarAffineFunction{Float64}-in-LessThan{Float64}
+ 0.0 + 1.0 v[1] - 1.0 v[2] <= 2.0
+
+VariableIndex-in-GreaterThan{Float64}
+ v[2] >= 0.0
+
+julia> map[c] isa MOI.ScalarAffineFunction{Float64}
+true
+```
+"""
+mutable struct PenaltyRelaxation{T}
+    default::Union{Nothing,T}
+    penalties::Dict{MOI.ConstraintIndex,T}
+
+    function PenaltyRelaxation(
+        p::Dict{MOI.ConstraintIndex,T};
+        default::Union{Nothing,T} = one(T),
+    ) where {T}
+        return new{T}(default, p)
+    end
+end
+
+function PenaltyRelaxation(; kwargs...)
+    return PenaltyRelaxation(Dict{MOI.ConstraintIndex,Float64}(); kwargs...)
+end
+
+function PenaltyRelaxation(
+    d::Dict{<:MOI.ConstraintIndex,T};
+    kwargs...,
+) where {T}
+    return PenaltyRelaxation(convert(Dict{MOI.ConstraintIndex,T}, d); kwargs...)
+end
+
+function MOI.modify(model::MOI.ModelLike, relax::PenaltyRelaxation{T}) where {T}
+    map = Dict{MOI.ConstraintIndex,MOI.ScalarAffineFunction{T}}()
+    for (F, S) in MOI.get(model, MOI.ListOfConstraintTypesPresent())
+        _modify_penalty_relaxation(map, model, relax, F, S)
+    end
+    return map
+end
+
+function _modify_penalty_relaxation(
+    map::Dict{MOI.ConstraintIndex,MOI.ScalarAffineFunction{T}},
+    model::MOI.ModelLike,
+    relax::PenaltyRelaxation,
+    ::Type{F},
+    ::Type{S},
+) where {T,F,S}
+    for ci in MOI.get(model, MOI.ListOfConstraintIndices{F,S}())
+        penalty = get(relax.penalties, ci, relax.default)
+        if penalty === nothing
+            continue
+        end
+        attr = ScalarPenaltyRelaxation(penalty)
+        if !MOI.supports(model, attr, MOI.ConstraintIndex{F,S})
+            @warn("Skipping PenaltyRelaxation of constraints of type $F-in-$S")
```

Comment:
`supports` is only defined for attributes and submittable. Is that `if` important ?